internal/exec/exec_test.go

```diff
@@ -0,0 +1,278 @@
+	}
+
+	procHeap, err := winapi.GetProcessHeap()
+	if err != nil {
+		t.Fatalf("GetProcessHeap failed: %s", err)
+	}
+
+	attrList, err := winapi.HeapAlloc(procHeap, 0, procThreadSize)
+	if err != nil {
+		t.Fatalf("HeapAlloc failed: %v\n", err)
+	}
+
+	procAttrList := (uintptr)(unsafe.Pointer(attrList))
+	if err = winapi.InitializeProcThreadAttributeList(procAttrList, num, 0, &procThreadSize); err != nil {
+		t.Fatalf("InitializeProcThreadAttributeList - second call failed: %v\n", err)
+	}
+
+	return procAttrList, procHeap, attrList
+}
+
+// Tests that the standard os/exec.Cmd functionality works the same.
+func TestExecProcess(t *testing.T) {
+	cmd := Command("ping", "127.0.0.1")
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("failed to run process: %s", err)
+	}
+}
+
+// Tests that the new procattrlist functionality works as expected.
+func TestExecProcessAttrList(t *testing.T) {
+	procAttrList, procHeap, attrList := createProcAttrList(t, 1)
+	defer winapi.HeapFree(procHeap, 0, attrList)
+	defer winapi.DeleteProcThreadAttributeList(procAttrList)
+
+	// This is useless but it tests all the win32 calls working. A better use for this test would be creating
+	// this process to be a child of a different process other than the one launching it but we can't gurantee
+	// any other processes running/get their pids.
+	pHandle, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, false, uint32(os.Getpid()))
+	if err != nil {
+		t.Fatalf("failed to open process: %s", err)
+	}
+	uintPtrHandle := uintptr(pHandle)
+	if err := winapi.UpdateProcThreadAttribute(
+		procAttrList,
+		0,
+		winapi.PROC_THREAD_ATTRIBUTE_PARENT_PROCESS,
+		&uintPtrHandle,
+		unsafe.Sizeof(pHandle),
+		nil,
+		nil,
+	); err != nil {
+		t.Fatalf("UpdateProcThreadAttribute failed: %s", err)
+	}
+
+	cmd := Command("ping", "127.0.0.1")
+	cmd.SysProcAttr = &SysProcAttr{
+		ProcThreadAttrList: procAttrList,
+	}
+
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("failed to run process: %s", err)
+	}
+}
+
+func TestJobObject(t *testing.T) {
+	procAttrList, procHeap, attrList := createProcAttrList(t, 1)
+	defer winapi.HeapFree(procHeap, 0, attrList)
+	defer winapi.DeleteProcThreadAttributeList(procAttrList)
+
+	job, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr("test"))
+	if err != nil {
+		t.Fatalf("failed to create job object")
+	}
+	uintPtrHandle := uintptr(job)
+	if err := winapi.UpdateProcThreadAttribute(
+		procAttrList,
+		0,
+		winapi.PROC_THREAD_ATTRIBUTE_JOB_LIST,
+		&uintPtrHandle,
+		unsafe.Sizeof(job),
+		nil,
+		nil,
+	); err != nil {
+		t.Fatalf("UpdateProcThreadAttribute failed: %s", err)
+	}
+
+	cmd := Command("ping", "127.0.0.1")
+	cmd.SysProcAttr = &SysProcAttr{
+		ProcThreadAttrList: procAttrList,
+	}
+
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("failed to run process: %s", err)
+	}
+}
+
+func TestPseudoConsole(t *testing.T) {
+	// Skip if not on RS5. This is when the pseudoconsole API was introduced
+	if osversion.Get().Build < osversion.RS5 {
```

Comment:
Checking OS version won't return correct values unless the application is properly manifested. You need to import `_ "github.com/Microsoft/hcsshim/test/functional/manifest"`.